astral-sh/ruff-vscode
# Pre-release for new `ruff server`-based extension

## Summary

This is a pre-release edition of Ruff for VS Code that uses our new, Rust-based LSP - `ruff server`.

At the moment, the pre-release has the following known limitations:

- Most extension settings are not yet supported, and many will not be needed after the transition to the new server is finished. For example, `editor.codeActionsOnSave` does not work at the moment due to missing support for source-level code actions (see below). Additionally, `lint.args` / `format.args` will be replaced in the future with specific configuration fields for the linter and formatter.
- Commands like `Fix all` and `Organize Imports` have not yet been implemented. (Quick Fixes should still work, though)
- Hierarchial configuration for individual files is not yet supported. At the moment, the language server uses the `ruff.toml` / `pyproject.toml` at the workspace root to configure the formatter and linter.
- Jupyter Notebook files are not supported yet.
- Using local Ruff binaries is not yet supported. At the moment, the extension will always use the bundled Ruff binary. (version `0.3.3`)

These limitations will be addressed in future versions.

To use the new Rust-based language server, you'll need to enable the "Experimental Server" setting and reload the extension. The setting can be disabled again if you encounter any major issues with the new server. 

## Test Plan

To test the features of the new extension, please do the following:

1. Enable the "Experimental Server" setting on the extension settings page and restart VS Code.

### Testing the Linter

2. Open a file in your repository of choice and write Python code that would normally cause a lint error in Ruff. For example, this code would cause three lint errors with the default rule selectors:
```python
def useless_semicolon():
    pass;

def multiple_fixes():
    x = True
    if x == False: print("this can't be good")
```
3. The lint errors should be highlighted in the editor view, and hovering over the highlighted text should display the correct diagnostic.
4. Select a portion of highlighted text and pull up a list of Quick Fixes with `Ctrl/Cmd + .`, or whatever shortcut you use for Quick Fixes. The appropriate fix should appear in the list.
5. Select the fix. The file should be applied correctly.
6. Undo the fix with `Ctrl/Cmd+Z`.
7. Select the entire file with `Ctrl/Cmd+A`.
8. Open up the list of quick fixes for the file with `Ctrl/Cmd + .`. All the lint fixes should be in the dropdown list.
9. Apply each fix to the file, re-opening the list as necessary.
10. Confirm that the file passes the linter by running `ruff check <path to file>` from the workspace root folder.

### Testing the Formatter

11. Add miscellaneous whitespace throughout the file, such that it's incorrectly formatted in multiple places.
12. Highlight a section of incorrectly formatted code and run `Format Selection` from the right-click context menu or with the shortcut `Ctrl/Cmd+K+F`.
13. Confirm that the non-selected code was not edited, and that the currently selected code is now correctly formatted.
14. Undo this range format and test `Format Selection` by highlighting the entire file with `Ctrl/Cmd+A`.
16. Confirm that the file is formatted correctly by running `ruff format --check <path to file>` from the workspace root folder.
17. Undo the format change.
18. Right click and select `Format Document` from the VS Code context menu, or use the shortcut `Alt+Shift+F`.
19. Confirm that the file is formatted correctly by running `ruff format --check <path to file>` from the workspace root folder.

### Testing the Configuration Reloader
20. To test that your workspace configuration reloads, open the `ruff.toml`/`pyproject.toml` configuration file at your workspace root (if one exists). If it does not exist, create one.
21. Decide on a rule to test (ideally one that is not already selected in the config)
22. Open a Python file within the workspace and write code that would violate this rule. Confirm that it does not raise a lint error.
23. Open the config file and add this rule in `select`.
24. At the moment, diagnostics can be a bit stale after a configuration reload (this will be fixed in a future release). To get around this, edit a single character of the offending code to 'refresh' it.
25. The lint error for the rule should now appear on the offending code.
26. Remove the rule from `select` in the config.
27. 'Refresh' the offending code, and confirm that the lint error no longer appears.
